public/app/plugins/datasource/cloudwatch/components/MetricStatEditor/MetricStatEditor.tsx

```diff
@@ -32,6 +33,24 @@ export function MetricStatEditor({
     onRunQuery();
   };
 
+  const onNamespaceChange = async (query: CloudWatchMetricsQuery) => {
+    const validatedQuery = await validateMetricName(query);
+    onQueryChange(validatedQuery);
+  };
+
+  const validateMetricName = async (query: CloudWatchMetricsQuery) => {
+    let { metricName, namespace, region } = query;
+    if (!metricName) {
+      return query;
+    }
+    await datasource.getMetrics(namespace, region).then((result: Array<SelectableValue<string>>) => {
```

Comment:
Here we should be able to use the `metrics` from line 28 so we don't have to load the metrics from the backend.

Reply:
Thanks, 
actually the `metrics` line 28 represent the metrics for current `namespace` (before `onChange`) and I need the future metrics (for selected namespace, but component has not re-rendered yet, props are still the old ones).
I could not find a nicer way to do it though...

Reply:
You can move the `metricName` removal to a `useEffect` hook that only executes when `metrics` change (so we don't request metrics twice).

Reply:
Ah got it, makes sense! These metrics are hard coded in the backend anyway so it's okay.

Reply:
> You can move the metricName removal to a useEffect hook that only executes when metrics change (so we don't request metrics twice).

Do you mean when `namespace` changes? It's already in `onNamespaceChange` so should only be called then.
Also, I wanted the check to happen ***before*** re-render (that way the parents - and state - always have a valid query) so the component props won't have changed yet, I don't think I could use a hook then 🤔  ?

Reply:
no, I mean when `metrics` change (and removing the check in `onNamespaceChange` to avoid the double call).

> Also, I wanted the check to happen before re-render (that way the parents - and state - always have a valid query) so the component props won't have changed yet, I don't think I could use a hook then  ?

It's generally fine to have a temporary "wrong" query (for example, you have it when you are building the query and some fields are missing). What we want to avoid is to actually perform the query request if the data is not correct. Because of that, when the namespace change, I would use `onChange` from the props rather than  the local `onQueryChange` since the second function runs the query. We don't want to run it yet if the metric is not set.

I slightly prefer not to do the `getMetrics` call twice so I would do the hook thing but since it's a relatively cheap API call (because the hardcoded response in the backend) both solutions are fine, you can go with the one you prefer.

Reply:
I understand now! (I read as `metricName` before). 
I'm pretty much 50/50 split on what the best but if it's cheap enough, I think I would stick with my first approach to:
* Validate before update rather than after - thus keeping what's on the screen consistent with the state
* For readability, I tried the hook way but then I have to store the validated metricName (as modifying props is an anti-pattern) and add it to `query` when calling `onChange`: it might seem puzzling to read.